### ISSUE TITLE
TraceQL: Sorting traces when searching from Grafana

### ISSUE THREAD

**Issue by actionbronson** [grafana/tempo#2659]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by joe-elliott** [comment]
Howdy, thanks for the request.

> I'm unsure if this is a Grafana problem at the presentation layer or a Tempo problem

I suppose it depends on what the real need is here. Currently Tempo's search results are non deterministic. If you choose a 24 hour window it will return the first N results that happen to match the query. 

If you want to be able to explicitly retrieve the first N results in the specified time range in time order then we would need to make changes to deterministically search the Tempo backend.

If you want to just have a sortable table in Grafana with the results you are getting back then this would be a frontend/Grafana feature request.

WDYT?

**Comment by actionbronson** [comment]
Hey Joe,

	Thanks for the reply.  Ideally it would be to be able to have Tempo sort and return the results with an order-by field, e.g. an intrinsic field such as duration.  An example use-case is to query traces based on a few fields and return the Top 10 worst performers.  Right now we can do something like {.field = “ABC” && duration > 10m} which typically returns our worst-cases but this is based on a few heuristics.

It’d be awesome to be able to {.field=“ABC” && duration > 10m} | sort(duration).

**Comment by joe-elliott** [comment]
So this would be difficult to do generically, but not impossible. For a given time range we would always have to do a completely exhaustive search and then pass all results up to the query frontend to do the sorting and return the requested number of traces.

Specifically start time would be a deterministic version of the search that always returned the N most recent traces which we have considered in the past. This would be a bit easier b/c we could use the block time ranges to not perform a completely exhaustive search.

I'm not opposed to this feature, and will tag it so it stays on our radar, but I do think this is a back burner item.

**Comment by swar8080** [comment]
+1 for the ability to sort by duration

We have a dashboard to explore latency issues for a specific span, so it'd be nice to display the longest duration traces in the dashboard's time range

**Comment by ADavis11** [comment]
> We have a dashboard to explore latency issues for a specific span, so it'd be nice to display the longest duration traces in the dashboard's time range


We have the same thing in our env and would love an option to get the longest durations.

**Comment by peterdk** [comment]
Same issue here, I notice that my overview of the last hour Tempo traces in Grafana doesn't display latest results, but apparently random results from the selected timeframe (1 hour). I would prefer a `order by time desc` or similar solution in the TraceQL query.

**Comment by Chaofan1014** [comment]
I am currently facing the same issue. It would be great if TraceQL supported sorting.

**Comment by Dexiion** [comment]
Hi @joe-elliott  we are trying to use the most_recent query hint but It is not working, are we doing something wrong?

![Image](https://github.com/user-attachments/assets/177c05b0-7588-4c3a-87f7-499a99aaf834)

The query was made at 10:10 and the first trace It shows is from 09:43 but there are more recent ones as you can see on the image

![Image](https://github.com/user-attachments/assets/f376a405-2af9-41f1-aa1e-f9e382d0b035)

**Comment by joe-elliott** [comment]
I am unfortunately not able to repro this. Running a query like:

```
{ resource.service.name = "tempo-querier" } with (most_recent=true)
```

and changing the time range from last hour, last 30 mins, last 5 mins, etc renders consistent results. When new results come in the old ones are pushed down as expected. If you can repro this consistently can you start a new issue and give as much detail as possible?

**Comment by sam-mcbr** [comment]
@joe-elliott has the `with (most_recent=true)` query hint released in a version of tempo? We are running Tempo 2.7.2 (latest release currently) and Grafana 11.6.0 and I'm not seeing consistent sets of traces returned.

For example, here I have a static, absolute time range selected and if I keep rerunning this query (`{} with (most_recent=true)`) I see different results returned:

<img width="630" alt="Image" src="https://github.com/user-attachments/assets/eacbd861-17d4-4712-927d-3bbea669a702" />

<img width="648" alt="Image" src="https://github.com/user-attachments/assets/60e32206-6329-4d4f-9217-9204251f11a6" />

I ask about this being released because in the CHANGELOG, I see the feature is still listed under the `main / unreleased` section.

Am really excited for the option!

**Comment by joe-elliott** [comment]
Yup, `with (most_recent=true)` will exist in 2.8. We have a longer term goal of making this default behavior and for most scales this will return quickly. Unfortunately for our largest scale clusters making this behavior default was too much of regression.

**Comment by Hu1buerger** [comment]
most_recent is an amazing improvent! But searching for the 'worst performers' [i.e. as described here](https://github.com/grafana/tempo/issues/2659#issuecomment-1638821026) seems to be important as well

**Comment by amgjum** [comment]
> Hi [@joe-elliott](https://github.com/joe-elliott) we are trying to use the most_recent query hint but It is not working, are we doing something wrong?嗨，我们正在尝试使用 most_recent 查询提示，但它不起作用，我们是不是做错了什么？
> 
> ![Image](https://github.com/user-attachments/assets/177c05b0-7588-4c3a-87f7-499a99aaf834)
> 
> The query was made at 10:10 and the first trace It shows is from 09:43 but there are more recent ones as you can see on the image查询是在 10：10 进行的，它显示的第一个跟踪是从 09：43 开始的，但正如您在图像上看到的那样，还有更新的跟踪
> 
> ![Image](https://github.com/user-attachments/assets/f376a405-2af9-41f1-aa1e-f9e382d0b035)

hello ,I face this problem, but I  do not know how to Resolve the issue where the data is different every time.

**Comment by amgjum** [comment]
> I am unfortunately not able to repro this. Running a query like:不幸的是，我无法重现这一点。运行如下查询：
> 
> ```
> { resource.service.name = "tempo-querier" } with (most_recent=true)
> ```
> 
> and changing the time range from last hour, last 30 mins, last 5 mins, etc renders consistent results. When new results come in the old ones are pushed down as expected. If you can repro this consistently can you start a new issue and give as much detail as possible?更改时间范围从过去一小时、过去 30 分钟、过去 5 分钟等可呈现一致的结果。当新结果出现时，旧结果会按预期下推。如果您可以始终如一地重现这一点，您能否开始一个新问题并提供尽可能多的细节？

hello，I also face this issue。 same query in one time。but backend respnse different data about trace in  tempo datasource and drilldown trace。 

gafana 12，
tempo 2.7 microservices deployment mode

**Comment by joe-elliott** [comment]
`most_recent=true` will be in Tempo 2.8

**Comment by QuinnBast** [comment]
Great to hear! Sad that this is not available in older versions of Tempo already, as I'm using Tempo already and upgrading will require convincing our customer to do so, but this gives us a reason I suppose. But not having this by default makes finding traces extremely difficult since they are random...
